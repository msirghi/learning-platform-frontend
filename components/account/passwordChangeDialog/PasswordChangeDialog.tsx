import {
  Button,
  Dialog,
  DialogActions,
  DialogContent,
  DialogContentText,
  DialogTitle,
  TextField
} from '@material-ui/core';
import React, { useEffect, useState } from 'react';
import { AlertType } from '../../../common/enums';
import { useTranslation } from '../../../i18n';
import { AlertMessage } from '../../common/AlertMesssage';
import styles from './PasswordChangeDialog.module.scss';

type Props = {
  open: boolean;
  toggleOpen: (val: boolean) => void;
};

export const PasswordChangeDialog: React.FC<Props> = ({ open, toggleOpen }) => {
  const [oldPassword, setOldPassword] = useState('');
  const [newPassword, setNewPassword] = useState('');
  const [repeatPassword, setRepeatPassword] = useState('');
  const [error, setError] = useState<string | undefined>();
  const { t } = useTranslation();

  const onSubmit = () => {
    setError(undefined);

    if (newPassword !== repeatPassword) {
      setError(t('account:passwordMismatch'));
      return;
    }

    // Backend call
  };

  useEffect(() => {
    setOldPassword('');
    setNewPassword('');
    setRepeatPassword('');
    setError(undefined);
  }, [open]);

  return (
    <Dialog open={open} onClose={() => toggleOpen(false)}>
      <DialogTitle>{t('account:passwordChangeDialogTitle')}</DialogTitle>
      <DialogContent>
        <DialogContentText>{t('account:passwordChangeDialogDescription')}</DialogContentText>
        {error && (
          <div className={styles.alert} data-testid='alert'>
            <AlertMessage type={AlertType.ERROR} message={error} />
          </div>
        )}
        <TextField
          label={t('account:oldPassword')}
          fullWidth
          variant='outlined'
          value={oldPassword}
          type='password'
          onChange={(e) => setOldPassword(e.target.value)}
          inputProps={{
            'data-testid': 'old-pass-field'
          }}
        />
        <div className={styles.field}>
          <TextField
            label={t('account:newPassword')}
            fullWidth
            variant='outlined'
            type='password'
            value={newPassword}
            onChange={(e) => setNewPassword(e.target.value)}
            inputProps={{
              'data-testid': 'new-pass-field'
            }}
          />
        </div>

        <div className={styles.field}>
          <TextField
            type='password'
            label={t('account:repeatNewPassword')}
            fullWidth
            variant='outlined'
            value={repeatPassword}
            onChange={(e) => setRepeatPassword(e.target.value)}
            inputProps={{
              'data-testid': 'repeat-pass-field'
            }}
          />
        </div>
      </DialogContent>
      <DialogActions>
        <Button onClick={() => toggleOpen(false)} color='primary'>
          {t('account:cancelButtonLabel')}
        </Button>
        <Button
          data-testid='submit-btn'
          onClick={() => onSubmit()}
          color='primary'
          disabled={!oldPassword || !newPassword || !repeatPassword}
        >
          {t('account:submitButtonLabel')}
        </Button>
      </DialogActions>
    </Dialog>
  );
};
