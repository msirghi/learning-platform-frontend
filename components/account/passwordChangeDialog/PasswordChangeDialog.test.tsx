import { Button, Dialog } from '@material-ui/core';
import { fireEvent, render } from '@testing-library/react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import React from 'react';
import { PasswordChangeDialog } from './PasswordChangeDialog';

describe('PasswordChangeDialog component', () => {
  const props = {
    open: true,
    toggleOpen: jest.fn()
  };

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should match the snapshot', () => {
    const wrapper = shallow(<PasswordChangeDialog {...props} />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should trigger prop method on onClose Dialog method', () => {
    const wrapper = shallow(<PasswordChangeDialog {...props} />);
    const dialog = wrapper.find(Dialog);
    expect(dialog).toHaveLength(1);

    dialog.props().onClose({} as any, 'escapeKeyDown');
    expect(props.toggleOpen).toBeCalledTimes(1);
  });

  it('should trigger prop method on cancel button click', () => {
    const wrapper = shallow(<PasswordChangeDialog {...props} />);
    const cancelButton = wrapper.find(Button).at(0);
    expect(cancelButton).toBeDefined();
    cancelButton.simulate('click');
    expect(props.toggleOpen).toBeCalledTimes(1);
  });

  it('should should have submit button disabled by default', () => {
    const wrapper = shallow(<PasswordChangeDialog {...props} />);
    const submitButton = wrapper.find(Button).at(1);
    expect(submitButton).toBeDefined();
    expect(submitButton.props().disabled).toBeTruthy();
  });

  it('should have disabled button if all fields are filled and old password is empty', () => {
    const { queryByTestId } = render(<PasswordChangeDialog {...props} />);

    const oldPassword = queryByTestId('old-pass-field');
    expect(oldPassword).toBeInTheDocument();

    const newPassword = queryByTestId('new-pass-field');
    expect(newPassword).toBeInTheDocument();

    const repeatPassword = queryByTestId('repeat-pass-field');
    expect(repeatPassword).toBeInTheDocument();

    fireEvent.change(newPassword, { target: { value: 'test' } });
    fireEvent.change(repeatPassword, { target: { value: 'test' } });

    const submitButton = queryByTestId('submit-btn');
    expect(submitButton).toBeInTheDocument();
    expect(submitButton).toBeDisabled();
  });

  it('should have disabled button if all fields are filled and new password is empty', () => {
    const { queryByTestId } = render(<PasswordChangeDialog {...props} />);

    const oldPassword = queryByTestId('old-pass-field');
    expect(oldPassword).toBeInTheDocument();

    const newPassword = queryByTestId('new-pass-field');
    expect(newPassword).toBeInTheDocument();

    const repeatPassword = queryByTestId('repeat-pass-field');
    expect(repeatPassword).toBeInTheDocument();

    fireEvent.change(oldPassword, { target: { value: 'test' } });
    fireEvent.change(repeatPassword, { target: { value: 'test' } });

    const submitButton = queryByTestId('submit-btn');
    expect(submitButton).toBeInTheDocument();
    expect(submitButton).toBeDisabled();
  });

  it('should have disabled button if all fields are filled and repeat new password is empty', () => {
    const { queryByTestId } = render(<PasswordChangeDialog {...props} />);

    const oldPassword = queryByTestId('old-pass-field');
    expect(oldPassword).toBeInTheDocument();

    const newPassword = queryByTestId('new-pass-field');
    expect(newPassword).toBeInTheDocument();

    const repeatPassword = queryByTestId('repeat-pass-field');
    expect(repeatPassword).toBeInTheDocument();

    fireEvent.change(oldPassword, { target: { value: 'test' } });
    fireEvent.change(newPassword, { target: { value: 'test' } });

    const submitButton = queryByTestId('submit-btn');
    expect(submitButton).toBeInTheDocument();
    expect(submitButton).toBeDisabled();
  });

  it('should have enabled button if all fields are filled', () => {
    const { queryByTestId } = render(<PasswordChangeDialog {...props} />);

    const oldPassword = queryByTestId('old-pass-field');
    const newPassword = queryByTestId('new-pass-field');
    const repeatPassword = queryByTestId('repeat-pass-field');

    fireEvent.change(oldPassword, { target: { value: 'test' } });
    fireEvent.change(newPassword, { target: { value: 'test' } });
    fireEvent.change(repeatPassword, { target: { value: 'test' } });

    const submitButton = queryByTestId('submit-btn');
    expect(submitButton).toBeInTheDocument();
    expect(submitButton).not.toBeDisabled();
  });

  it('should should show the error message if passwords do not match', () => {
    const { queryByTestId } = render(<PasswordChangeDialog {...props} />);

    const oldPassword = queryByTestId('old-pass-field');
    const newPassword = queryByTestId('new-pass-field');
    const repeatPassword = queryByTestId('repeat-pass-field');

    fireEvent.change(oldPassword, { target: { value: 'test' } });
    fireEvent.change(newPassword, { target: { value: 'test' } });
    fireEvent.change(repeatPassword, { target: { value: 'testtest' } });

    const submitButton = queryByTestId('submit-btn');
    expect(submitButton).toBeInTheDocument();
    expect(submitButton).not.toBeDisabled();

    fireEvent.click(submitButton);

    const alert = queryByTestId('alert');
    expect(alert).toBeInTheDocument();
  });
});
