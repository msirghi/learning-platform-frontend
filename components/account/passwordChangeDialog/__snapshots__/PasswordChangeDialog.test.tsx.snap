// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PasswordChangeDialog component should match the snapshot 1`] = `
<WithStyles(ForwardRef(Dialog))
  onClose={[Function]}
  open={true}
>
  <WithStyles(ForwardRef(DialogTitle))>
    account:passwordChangeDialogTitle
  </WithStyles(ForwardRef(DialogTitle))>
  <WithStyles(ForwardRef(DialogContent))>
    <WithStyles(ForwardRef(DialogContentText))>
      account:passwordChangeDialogDescription
    </WithStyles(ForwardRef(DialogContentText))>
    <WithStyles(ForwardRef(TextField))
      fullWidth={true}
      inputProps={
        Object {
          "data-testid": "old-pass-field",
        }
      }
      label="account:oldPassword"
      onChange={[Function]}
      type="password"
      value=""
      variant="outlined"
    />
    <div
      className="field"
    >
      <WithStyles(ForwardRef(TextField))
        fullWidth={true}
        inputProps={
          Object {
            "data-testid": "new-pass-field",
          }
        }
        label="account:newPassword"
        onChange={[Function]}
        type="password"
        value=""
        variant="outlined"
      />
    </div>
    <div
      className="field"
    >
      <WithStyles(ForwardRef(TextField))
        fullWidth={true}
        inputProps={
          Object {
            "data-testid": "repeat-pass-field",
          }
        }
        label="account:repeatNewPassword"
        onChange={[Function]}
        type="password"
        value=""
        variant="outlined"
      />
    </div>
  </WithStyles(ForwardRef(DialogContent))>
  <WithStyles(ForwardRef(DialogActions))>
    <WithStyles(ForwardRef(Button))
      color="primary"
      onClick={[Function]}
    >
      account:cancelButtonLabel
    </WithStyles(ForwardRef(Button))>
    <WithStyles(ForwardRef(Button))
      color="primary"
      data-testid="submit-btn"
      disabled={true}
      onClick={[Function]}
    >
      account:submitButtonLabel
    </WithStyles(ForwardRef(Button))>
  </WithStyles(ForwardRef(DialogActions))>
</WithStyles(ForwardRef(Dialog))>
`;
